File: frontend/src/types/solicitud.ts
// Espejo de backend/app/schemas/solicitud.py (Regla 5)
import type { EstadoSolicitud } from "./common";
import type { UsuarioPublic } from "./usuario";

export interface ItemSolicitudCreate {
  producto_id: number;
  cantidad: number;
}

export interface SolicitudCreate {
  items: ItemSolicitudCreate[];
}

export interface DetalleSolicitudPublic {
  producto_id: number;
  nombre_producto: string;
  cantidad: number;
  precio_unitario: number;
}

export interface SolicitudPublic {
  id: number;
  usuario_id: number;
  estado: EstadoSolicitud;
  total: number;
  creado_en: string; // ISO 8601
  pagada_en: string | null; // ISO 8601 | null
  items: DetalleSolicitudPublic[];
}

export interface SolicitudAdminPublic extends SolicitudPublic {
  usuario: UsuarioPublic;
}

export interface SolicitudesResponse {
  items: SolicitudPublic[];
  page: number;
  page_size: number;
  total: number;
}

export interface SolicitudesAdminResponse {
  items: SolicitudAdminPublic[];
  page: number;
  page_size: number;
  total: number;
}

export interface PagoResponse {
  solicitud_id: number;
  estado: "pagada";
  total: number;
  pagada_en: string; // ISO 8601
  unidades_descontadas: number;
}

export interface FaltanteInfo {
  producto_id: number;
  nombre: string;
  solicitado: number;
  disponible: number;
}

export interface FaltantesResponse {
  detail: string;
  faltantes: FaltanteInfo[];
}

File: backend/app/schemas/solicitud.py
# -*- coding: utf-8 -*-
"""Schemas de solicitudes (ARQ-T2)."""
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import Field

from .auth import UsuarioPublic
from .common import ESTADO_SOLICITUD, SchemaBase
from .producto import _MontoJson


class ItemSolicitudCreate(SchemaBase):
    producto_id: int
    cantidad: int = Field(ge=1, le=999)


class SolicitudCreate(SchemaBase):
    items: list[ItemSolicitudCreate] = Field(min_length=1, max_length=50)


class DetalleSolicitudPublic(_MontoJson):
    producto_id: int
    nombre_producto: str
    cantidad: int
    precio_unitario: Decimal


class SolicitudPublic(_MontoJson):
    id: int
    usuario_id: int
    estado: ESTADO_SOLICITUD
    total: Decimal
    creado_en: datetime
    pagada_en: datetime | None
    items: list[DetalleSolicitudPublic]


class SolicitudAdminPublic(SolicitudPublic):
    """Vista de administración — agrega los datos del usuario."""

    usuario: UsuarioPublic


class SolicitudesResponse(SchemaBase):
    items: list[SolicitudPublic]
    page: int
    page_size: int
    total: int


class SolicitudesAdminResponse(SchemaBase):
    items: list[SolicitudAdminPublic]
    page: int
    page_size: int
    total: int


class PagoResponse(_MontoJson):
    """Resultado de POST /api/admin/solicitudes/{id}/pagar."""

    solicitud_id: int
    estado: Literal["pagada"]
    total: Decimal
    pagada_en: datetime
    unidades_descontadas: int


class FaltanteInfo(SchemaBase):
    """Producto con stock insuficiente al intentar pagar (HTTP 409)."""

    producto_id: int
    nombre: str
    solicitado: int
    disponible: int


class FaltantesResponse(SchemaBase):
    """Cuerpo del error 409: detalle + faltantes (no se descuenta nada parcial)."""

    detail: str
    faltantes: list[FaltanteInfo]
